MSC4108: Mechanism to allow OAuth 2.0 API sign in and E2EE set up via QR code (#4108)
matrix-org/matrix-doc: proposals/4108-oidc-qr-login.md

Reviewer: @mscbot concern Dependency on MSC4388 outstanding

Regardless of which device generates the QR code, either device can be the existing (already signed in) device. The
other device is then the new device (one seeking to be signed in).

Symmetric encryption uses deterministic nonces, incrementing by `2` with each payload. Device S starts with `0`, using only
even nonces. Device A starts with `1`, using only odd nonces.

1. **Ephemeral key pair generation**

  Both devices generate an _ephemeral_ Curve25519 key pair:

- Device G generates **(Gp, Gs)**, where **Gp** is its public key and **Gs** the private (secret) key.
- Device S generates **(Sp, Ss)**, where **Sp** is its public key and **Ss** the private (secret) key.

2. **Create rendezvous session**

Device G creates a rendezvous session by making a `POST` request (as described previously) to the nominated homeserver
with an empty payload. It parses the **id** and **server** received.

3. **Initial key exchange**

Device G displays a QR code containing:

- its public key **Gp**
- the insecure rendezvous session **URL**
- An indicator (the **intent**) to say if this is a new device which wishes to "initiate" a login, or an existing device

> that wishes to "reciprocate" a login

Reviewer 2: A better idea is probably to just include the intent in the HKDF info, making this check implicit.

Mostly commenting as a note to myself to modify this, but discussion is welcome.

Reviewer 3: That would result in a worse error message. Not saying that we shouldn't remove it, just that the intent might have some usage.

Author: The reason for putting the intent in the QR is so that the Matrix client that has scanned can immediately determine if the QR code is usable for whatever the user has asked it to do.

Then, as @poljar suggests, if it is the wrong type then it can display a suitable error message.

In the QR code representation it is encoded in the existing single `mode` byte. So, removing the intent from the QR code would not reduce the size.

> that wishes to "reciprocate" a login

Author: I agree that the terms are not ideal. I think I chose them as they are used elsewhere in the Client-Server API specification. I'm not sure "petition" and "grant" are better though!

Another way we could describe it is by defining the disposition of the device.

So, the "initiator" could become the "un-authenticated device". And the "reciprocator" could become the "authenticated device"?

> that wishes to "reciprocate" a login

Reviewer 4: I think a device, the unauthenticated one, "wishes" to login and the other, authenticated, one "grants" it.

> that wishes to "reciprocate" a login

Author: > I think a device, the unauthenticated one, "wishes" to login and the other, authenticated, one "grants" it.

This "grants" terminology clashes with OIDC.

In OIDC the OIDC Provider "grants" access after the user has "consented".

Specifically, the device doesn't have the authority to grant access. Only the OIDC Provider.

> that wishes to "reciprocate" a login

Reviewer 4: What about  `which wishes to login, or an existing device that wishes to open for a login" then?

they are both in the correct state to the user performing the QR code login process.

#### Establishment

Participants:

- Device G (the device generating the QR code)
- Device S (the device scanning the QR code)

Regardless of which device generates the QR code, either device can be the existing (already signed in) device. The
other device is then the new device (one seeking to be signed in).

Symmetric encryption uses a separate encryption key for each sender, both derived from a shared secret using HKDF. A
separate deterministic, monotonically-incrementing nonce is used for each sender. Devices initially set both nonces to
`0` and increment the corresponding nonce by `1` for each message sent and received.

1. **Ephemeral key pair generation**

  Both devices generate an _ephemeral_ Curve25519 key pair:

- Device G generates **(Gp, Gs)**, where **Gp** is its public key and **Gs** the private (secret) key.
- Device S generates **(Sp, Ss)**, where **Sp** is its public key and **Ss** the private (secret) key.

2. **Create rendezvous session**

Device G creates a rendezvous session by making a `POST` request (as described previously) to the nominated homeserver

> with an empty payload. It parses the **id** and **server** received.

Reviewer 5: Unless I am having trouble following this, isn't the response a JSON object with a `url`?

Author: You are quite right. This should be talking about the `url`.

Fixed in 25e8fcb2c34089efd707a98917d24a57d26d23b9.

just pressing "continue", and that the Device S has been able to determine that the channel is secure.

Device G compares the code that the user has entered with the **CheckCode** that it calculates using the same mechanism
as before:

```
CheckBytes := HKDF_SHA256(SH, "MATRIX_QR_CODE_LOGIN_CHECKCODE|" || Gp "|" || Sp , salt=0, size=2)
CheckCode := NumToString(CheckBytes[0] % 10) || NumToString(CheckBytes[1] % 10)
```

If the code that the user enters matches then the secure channel is established.

Subsequent payloads sent from G should be encrypted using **EncKey_G**, while payloads sent from S should be
encrypted with **EncKey_S**, incrementing the corresponding nonce for each message sent/received.

#### Sequence diagram

The sequence diagram for the above is as follows:

```mermaid
sequenceDiagram
    participant G as Device G
    participant Z as Rendezvous server
    participant S as Device S

    note over G,S: 1) Devices G and S each generate an ephemeral Curve25519 key pair

    activate G
    note over G: 2) Device G creates a rendezvous session as follows
    G->>+Z: POST /_matrix/client/rendezvous

>     Z->>-G: 201 Created<br>Location: /_matrix/client/rendezvous/abc-def<br>ETag: 1

Reviewer 5: Here the location is in the JSON body, not in the `LOCATION` header, isn't it?

Author: Again, well spotted. This is fixed in e12945c5fdb9b831edc4f702862cdb433a96961b along with some other inconsistencies.

- `202 Accepted` - payload updated
- `400 Bad Request` (`M_MISSING_PARAM`) - a required header was not provided.
- `400 Bad Request` (`M_INVALID_PARAM`) - a malformed 
[`ETag`](https://github.com/matrix-org/matrix-spec-proposals/blob/hughns/simple-rendezvous-capability/proposals/3886-simple-rendezvous-capability.md#the-update-mechanism)
header was provided or invalid `Content-Type`.
- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session URL is not valid (it could have expired)
- `412 Precondition Failed` (`M_CONCURRENT_WRITE`, a new error code) - when the ETag does not match
- `413 Payload Too Large` (`M_TOO_LARGE`) - the supplied payload is too large
- `429 Too Many Requests` (`M_UNKNOWN`) - the request has been rate limited

HTTP response headers for `202 Accepted` and `412 Precondition Failed`:

- common headers as defined above

##### Receive a payload from the rendezvous session: `GET <rendezvous session URL>`

HTTP request headers:

- `If-None-Match` - optional, as per [RFC7232](https://httpwg.org/specs/rfc7232.html#header.if-none-match) server will
only return data if given ETag does not match

HTTP response codes, and Matrix error codes:

- `200 OK` - payload returned

> - `304 Not Modified` - when `If-None-Match` is supplied and the ETag does not match

Reviewer 5: This is actually if the ETag matches the one provided, right?

|`homeserver`|`string`| When the existing device is sending this it can include the Base URL of the homeserver so that the new device can at least save the user the hassle of typing it in|

Example:

```json
{
    "type":"m.login.failure",

>     "reason": "unsupported",

Reviewer 3: `unsupported`, according to the above table, doesn't exist.

```suggestion
    "reason": "unsupported_protocol",
```

    "homeserver": "https://matrix-client.matrix.org"
}
```

##### `m.login.declined`

> Sent by: existing device

Reviewer 3: ```suggestion
Sent by: new device
```

The graphs show that this is sent by the new device.

Author: Further to our conversation just now, I concur.

```
Nonce_G := 1
NonceBytes_G := ToLowEndianBytes(Nonce_G)[..12]
TaggedCiphertext := ChaCha20Poly1305_Encrypt(EncKey_G, NonceBytes_G, "MATRIX_QR_CODE_LOGIN_OK")
Nonce_G := Nonce_G + 1
LoginOkMessage := UnpaddedBase64Encode(TaggedCiphertext)
```

Device G sends **LoginOkMessage** as the payload via `PUT` request with `Content-Type` header set to `text/plain` to the
insecure rendezvous session.

6. **Verification by Device S**

Device S receives a response over the insecure rendezvous session by polling with `GET` requests, potentially from
Device G.

It decrypts (and authenticates) it using the previously computed encryption key, which will succeed provided the payload
was indeed sent by Device G. It then verifies the plaintext matches `MATRIX_QR_CODE_LOGIN_OK`, failing otherwise.

```
Nonce_G := 1
(TaggedCiphertext, Sp) := Unpack(Message)
NonceBytes := ToLowEndianBytes(Nonce)[..12]
Plaintext := ChaCha20Poly1305_Decrypt(EncKey_G, NonceBytes, TaggedCiphertext)
Nonce_G := Nonce_G + 1

unless Plaintext == "MATRIX_QR_CODE_LOGIN_OK":
     FAIL
```

If the above was successful, Device S then calculates a two digit **CheckCode** code derived from **SH**, **Gp** and
**Sp**:

```

> CheckBytes := HKDF_SHA256(SH, "MATRIX_QR_CODE_LOGIN_CHECKCODE|" || Gp "|" || Sp , salt=0, size=2)

Reviewer 2: ```suggestion
CheckBytes := HKDF_SHA256(SH, "MATRIX_QR_CODE_LOGIN_CHECKCODE|" || Gp || "|" || Sp , salt=0, size=2)
```

Example:

```json
{
    "type": "m.login.secrets",
    "cross_signing": {
        "master_key": "$base64_of_the_key",
        "self_signing_key": "$base64_of_the_key",
        "user_signing_key": "$base64_of_the_key"
    },
    "backup": {
        "algorithm": "foobar",
        "key": "$base64_of_the_backup_recovery_key",
        "backup_version": "version_string"
    }
}
```

### QR code format

The proposed format of the QR code intends to be similar to that which is already described in the Client-Server API for
[device verification](https://spec.matrix.org/v1.9/client-server-api/#qr-code-format).

Additional modes are added to the byte used for "QR code verification mode" to allow for the two login intents: initiate
on a new device; reciprocate on an existing device;

The QR codes to be displayed and scanned using this format will encode binary strings in the general form:

- the ASCII string `MATRIX`
- one byte indicating the QR code version (must be `0x02`)
- one byte indicating the QR code intent/mode. Should be one of the following values:

>   - `0x03` a new device wishing to initiate a login and self-verify

Reviewer 6: To my knowledge, none of the clients listed in the PR description support this new device mode. Should there be a client that implements this before merging to the spec?
(I commented on the wrong mode at first, this is confusing)

Reviewer 7: If there are no available implementation of `0x03`, then that seems like a significant omission, and it may block this MSC getting FCP'ed

#### Other existing protocols

One could try and do something with STUN or TURN or [COAP](https://datatracker.ietf.org/doc/html/rfc7252).

#### Implementation details

Rather than requiring the devices to poll for updates, "long-polling" could be used instead similar to `/sync`. Or WebSockets.

### Alternative method of secret sharing

Instead of the existing device sharing the secrets bundle instead the existing device could cross-sign the new device
and then use to-device messaging for sharing the secrets.

For:

- You re-use existing secret sharing

Against:

- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.
- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow
- The backup cannot be immediately enabled since we received the backup version as well, something the `m.secret.send`
mechanism does not offer.
- The new device cannot upload the cross-signing signature with the device keys in a single request. This introduces a
chance of other devices seeing the new device as unverified, incorrectly prompting the user to verify the device that
will soon be verified.

> ## Security considerations

Reviewer 8: I think its worth noting that QR login allows an attacker with temporary access to an unlocked client to extend that to ongoing access, i.e. if you leave your phone unlocked briefly someone could quickly use QR code login to login on their device.

It also makes it easier to get the private keys of the user from an unlocked client, as you can login with a new device, extract the keys from that, and logout again to cover your tracks.

Sophisticated attackers can today use specialist equipment to extract private keys and access tokens from the memory of a process. However, a) that is a much higher bar for attack, and b) cloning an access token will quickly be detected via refresh tokens. 

I think we should add some recommendations to help mitigate this, something like:
- Servers SHOULD send new device login notifications to other devices or email
- Native clients SHOULD gate QR code login behind some form of extra protection, e.g. biometrics on mobile apps. These should be minimally invasive though as otherwise it heavily erodes the benefit of using QR code login in the first place. We don't necessarily think this protection is worth while on web clients, as its trivial to access the devtools to extract the secrets directly and/or bypass any extra protections.

## Proposal

Depending on the pair of devices used, it may be preferable to scan the QR code on either the new or existing device,
based on the availability of a camera. As such, this proposal allows for the generation of the QR on either device.

In order for the new device to be fully set up, it needs to exchange information with an existing device such that:

- The new device knows which homeserver to use
- The existing device can facilitate the new device in getting an access token
- The existing device shares the secrets necessary to set up end-to-end encryption

This proposal is split into three parts:

1. An insecure rendezvous session API to allow the two devices to exchange the necessary data
2. A secure channel to protect the data exchanged over the rendezvous session
3. The OAuth login part and set up of E2EE

### Insecure rendezvous session

It is proposed that an HTTP-based protocol be used to establish an ephemeral bi-directional communication session over
which the two devices can exchange the necessary data. This session is described as "insecure" as it provides no
end-to-end confidentiality nor authenticity by itself---these are layered on top of it.

> #### High-level description

Reviewer: Editorial: Having a sequence diagram early in the proposal here would help a lot in understanding the flow.

header. After this point, any further attempts to query or update the payload MUST fail. The rendezvous session can be
manually expired with a `DELETE` call to the rendezvous session.

####  API

##### Common HTTP response headers

- `ETag` - required, ETag for the current payload at the rendezvous session as per [RFC7232](https://httpwg.org/specs/rfc7232.html#header.etag)
- `Expires` - required, the expiry time of the rendezvous as per [RFC7234](https://httpwg.org/specs/rfc7234.html#header.expires)
- `Last-Modified` - required, the last modified date of the payload as per [RFC7232](https://httpwg.org/specs/rfc7232.html#header.last-modified)
- `Cache-Control` - required, `no-store` as per [RFC7234](https://httpwg.org/specs/rfc7234.html#header.cache-control)
- `Pragma` - required, `no-cache` as per [RFC7234](https://httpwg.org/specs/rfc7234.html#header.pragma)

##### Create a rendezvous session and send initial payload: `POST /_matrix/client/v1/rendezvous`

This would be part of the Client-Server API.

HTTP request headers:

- `Content-Length` - required
- `Content-Type` - required, must be `text/plain`

HTTP request body:

> - any data up to maximum size allowed by the server

Reviewer: we should specify what this limit is rather than leaving it up to the implementation, otherwise clients will have to guess & check, or assume.

Likewise on other HTTP endpoints/methods.

HTTP response codes, and Matrix error codes:

- `201 Created` - rendezvous session created
- `400 Bad Request` (``M_MISSING_PARAM``) - either `Content-Length` and/or `Content-Type` was not provided.
- `400 Bad Request` (`M_INVALID_PARAM`) - an invalid `Content-Type` was given.

> - `403 Forbidden` (``M_FORBIDDEN``) - forbidden by server policy

Reviewer: is 403 how a server can opt-out of this? Because the Matrix specification requires servers to implement all endpoints, without an opt-out mechanism it's possible for *any* server to become abused for this channel, even if the two users talking have nothing to do with that server.

Being able to apply some restrictions, like only allowing logged-in users to call this endpoint, is essential.

- `413 Payload Too Large` (``M_TOO_LARGE``) - the supplied payload is too large
- `429 Too Many Requests` (``M_UNKNOWN``) - the request has been rate limited
- `307 Temporary Redirect` - if the request should be served from somewhere else specified in the `Location` response header

n.b. the [`307 Temporary Redirect`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307) response code has been
chosen explicitly for the behaviour of ensuring that the method and body will not change whilst the user-agent follows
the redirect. For this reason, no other `30x` response codes are allowed.

HTTP response headers for `201 Created`:

- `Content-Type`- required, application/json
- common headers as defined above

HTTP response body for `201 Created`:

- a JSON object with a single key `url` whose value is the absolute URL of the rendezvous session

Example response:

```http
HTTP 201 Created
Content-Type: application/json
ETag: VmbxF13QDusTgOCt8aoa0d2PQcnBOXeIxEqhw5aQ03o=
Expires: Wed, 07 Sep 2022 14:28:51 GMT
Last-Modified: Wed, 07 Sep 2022 14:27:51 GMT
Cache-Control: no-store
Pragma: no-cache

{
    "url": "http://example.org/abcdEFG12345"
}
```

> ##### Send a payload to the rendezvous session: `PUT <rendezvous session URL>`

Reviewer: I'm very worried about unauthenticated endpoints accepting data the server needs to persist indefinitely. Rate limits, size limits, and longevity limits need to be described by the MSC.

HTTP request headers:

- `Content-Length` - required
- `Content-Type` - required, must be `text/plain`
- `If-Match` - required. The ETag of the last payload seen by the requesting device.

HTTP request body:

- any data up to maximum size allowed by the server

HTTP response codes, and Matrix error codes:

- `202 Accepted` - payload updated
- `400 Bad Request` (`M_MISSING_PARAM`) - a required header was not provided.
- `400 Bad Request` (`M_INVALID_PARAM`) - a malformed 
[`ETag`](https://github.com/matrix-org/matrix-spec-proposals/blob/hughns/simple-rendezvous-capability/proposals/3886-simple-rendezvous-capability.md#the-update-mechanism)
header was provided or invalid `Content-Type`.
- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session URL is not valid (it could have expired)
- `412 Precondition Failed` (`M_CONCURRENT_WRITE`, a new error code) - when the ETag does not match
- `413 Payload Too Large` (`M_TOO_LARGE`) - the supplied payload is too large
- `429 Too Many Requests` (`M_UNKNOWN`) - the request has been rate limited

HTTP response headers for `202 Accepted` and `412 Precondition Failed`:

- common headers as defined above

> ##### Receive a payload from the rendezvous session: `GET <rendezvous session URL>`

Reviewer 7: Could the rendezvous server just require some header, maybe an `Authorization` header even if it uses a dummy value, to prevent someone from accidentally opening the URL?

> ##### Receive a payload from the rendezvous session: `GET <rendezvous session URL>`

Reviewer 2: I like uhoreg's proposal. It feels much nicer than collapsing everything into a single SOAP-y endpoint.

> ##### Receive a payload from the rendezvous session: `GET <rendezvous session URL>`

Author: Some thoughts:

- Compared to the original proposal (which allowed any content type) this current one is better in that it restricts to `application/json` and only the `data` string is open to abuse. However, a risk does reamin.
- Even with a `POST` it is simple enough to have a form hosted somewhere that a victim's browser could navigate to.
- Could we instead treat this as a browser content/security policy issue> For example we could have the server check the [Sec-Fetch-Mode](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Sec-Fetch-Mode) header and return some error if the mode is `navigate`. I am thinking of this akin to how CORS works where it is a reasonable starting point to assume that a victim has a well-behaved browser that is setting headers as expected.

Thoughts?

HTTP response headers for `304 Not Modified`:

- common headers as defined above

HTTP response body for `200 OK`::

- The payload last set for this rendezvous session, either via the creation POST request or a subsequent PUT request, up
to the maximum size allowed by the server.

Example responses:

```http
HTTP 200 OK
Content-Type: text/plain
ETag: VmbxF13QDusTgOCt8aoa0d2PQcnBOXeIxEqhw5aQ03o=
Expires: Wed, 07 Sep 2022 14:28:51 GMT
Last-Modified: Wed, 07 Sep 2022 14:27:51 GMT
Cache-Control: no-store
Pragma: no-cache

foo
```

```http
HTTP 304 Not Modified
ETag: VmbxF13QDusTgOCt8aoa0d2PQcnBOXeIxEqhw5aQ03o=
Expires: Wed, 07 Sep 2022 14:28:51 GMT
Last-Modified: Wed, 07 Sep 2022 14:27:51 GMT
Cache-Control: no-store
Pragma: no-cache
```

##### Cancel a rendezvous session: `DELETE <rendezvous session URL>`

HTTP response codes:

- `204 No Content` - rendezvous session cancelled
- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session URL is not valid (it could have expired)
- `429 Too Many Requests` (`M_UNKNOWN`) - the request has been rate limited

##### Authentication

These API endpoints do not require authentication because trust is established at the secure channel layer which is

> described later.

Reviewer 8: One thought for restricting what you can do with unauthenticated rendezvous channel is to change it to instead work something like:

1. Unauthenticated device (UD) creates a "redirect channel" on HS1 and sets that in the QR code.
2. The authenticated device (AD) creates a rendezvous channel on HS2.
3. HS2 POSTS to the redirect channel on HS1 with the homeserver and rendezvous channel ID. HS1 validates its from HS2.
4. HS1 returns the homeserver (HS2) and rendezvous channel ID to UD, who then uses that channel as normal.

This has the following properties:
1. It limits how much information can be persisted on an unauthenticated channel. We can severely restrict the size of the request ID for example.
2. An abuser must use a domain they own if they want to encode dodgy data in the rendezvous channel ID. We can then ban abusive domains.
3. An unauthenticated device can only receive information, rather than create a 2-way channel. Not sure that's at all useful thing to assert, but it is nonetheless a property.

I think this sort of flow would reduce potential abuse vectors, but equally makes things more complicated and may not be worth it.

> described later.

Author: I have done this in 9dab408720d89ad4d96dabc7fb3ba353f8072aea

1. Check that the Device Authorization Grant is available as per [MSC4341]

For a new device it would need to know the homeserver ahead of time in order to do these checks.

Additionally the new device needs to either have an existing (i.e. static) OAuth client registered with the homeserver
already, or the homeserver must support and allow dynamic client registration as described in the [spec](https://spec.matrix.org/v1.15/client-server-api/#client-registration).

The feature is also only available where a user has cross-signing set up and the existing device to be used has the
Master Signing Key, Self Signing Key and User Signing Key stored locally so that they can be shared with the new device.

## Potential issues

Because this is an entirely new set of functionality it should not cause issue with any existing Matrix functions or capabilities.

The proposed protocol requires the devices to have IP connectivity to the server which might not be the case in P2P scenarios.

## Alternatives

### Alternative to the rendezvous session protocol

> #### Send-to-Device messaging

Reviewer: This is a compelling alternative, honestly. It causes the caller to be authenticated, *and* accept the Terms and Conditions explicitly rather than implicitly (which is questionable as-is). It also means traditional safety tooling can be applied to the user. 

I think it's also fair to say that QR code login isn't available on servers which disable guest access. Single-user (or low-user) homeservers can enable it as-needed, and private homeservers can apply network conditions or registration requirements to disable guest access to the wider internet. Public homeservers are likely to already want/have guest access enabled.

Being able to "upgrade" a guest account to an existing account would be ideal as well, though for purposes of scoping can/should be deferred to yet another future MSC.

Reviewer 8: One thing that I think is not immediately obvious from this MSC is that there are two ways of starting a QR code session:
1. Generating a QR code on a logged in device. This device can create a rendezvous channel on its local HS, and so this can be authenticated. 
2. Generating a QR code on a fresh device. The client has no idea what account or HS its going to sign it to, as its the other device that is logged in (that will scan the QR code). The client then creates a rendezvous channel on a static pre-configured server (e.g. matrix.org), but with no authentication.

This has some ramifications:
1.  We can default to homeservers require authentication to *create* rendezvous channels, which helps T&S for the majority of HSes. 
2. Clients that want to generate QR codes when logged out must pre-configure a homeserver to accept unauthenticated rendezvous channels. This is where we need to be careful with T&S, but at least we've limited the scope and can potentially limit it further (maybe make it so that the rendezvous channel redirects to the actual HS or something)
3. I don't think that using to-device channels really help in this situation. To handle the case of non-logged in devices you'd need to allow automatic registration of a temporary user and device on a 3rd party HS (and we would strongly want to avoid having to manually accept terms for a 3rd party server from a product PoV), which just feels way more heavyweight than allowing locked down rendezvous channels.

I'm going to add a new comment about adding auth

Because the rendezvous session is not authenticated, it is possible for an attacker to use it to distribute malicious
content.

This could lead to a reputational problem for the homeserver domain or IPs, as well as potentially causing harm to users.

Mitigations that are included in this proposal:

- the low maximum payload size
- restricted allowed content type
- the rendezvous session should be short-lived

### Secure channel

The above rendezvous session is insecure, providing no confidentiality nor authenticity against the rendezvous server or
even arbitrary network participants which possess the rendezvous session ID and server name.
To provide a secure channel on top of this insecure rendezvous session transport, we propose the following scheme.

This scheme is essentially [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme#Formal_description_of_ECIES)
instantiated with X25519, HKDF-SHA256 for the KDF and ChaCha20-Poly1305 (as specified by
[RFC8439](https://datatracker.ietf.org/doc/html/rfc8439#section-2.8)) for the authenticated encryption. Therefore,
existing security analyses of ECIES are applicable in this setting too. Nevertheless we include below a short

> description of our instantiation of ECIES and discuss some potential pitfalls and attacks.

Reviewer 2: If we were designing this today, I'd use [RFC 9180 HPKE](https://datatracker.ietf.org/doc/rfc9180/) instead. It's conceptually the same construct as ECIES but better specified, backed by the CFRG/IRTF and used in MLS.

To get a good trade-off between visual compactness and high level of error correction we use a binary mode QR with a
similar structure to that of the existing Device Verification QR code encoding described in [Client-Server
API](https://spec.matrix.org/v1.9/client-server-api/#qr-code-format).

This is defined in detail in a separate section of this proposal.

Device S scans and parses the QR code to obtain **Gp**, the rendezvous session **URL**, **intent** and optionally the Matrix homeserver
**[server name](https://spec.matrix.org/v1.15/appendices/#server-name)**.

At this point Device S should check that the received intent matches what the user has asked to do on the device.

4. **Device S sends the initial payload**

Device S computes a shared secret **SH** by performing ECDH between **Ss** and **Gp**. It then discards **Ss** and
derives two 32-byte symmetric encryption keys from **SH** using HKDF-SHA256. One of those keys, **EncKey_S** is
used for messages encrypted by device S, while the other, **EncKey_G** is used for encryption by device G.

The keys are generated with the following HKDF parameters:

**EncKey_S**

> - `MATRIX_QR_CODE_LOGIN_ENCKEY_S|Gp|Sp` as the info, where **Gp** and **Sp** stand for the generating

Reviewer 7: I don't think any of this is specific to QR codes, and could be used by any other authentication mechanism that wants to use the rendezvous service similarly in the future.  (NFC or something?)  So I would suggest dropping the `QR_CODE` from all the strings.

Author: It was a deliberate choice to include the `QR_CODE` in the prefix so that the protocol is bound to the specific application (the QR code login).

Whilst the could be some future usage of the secure channel scheme (e.g NFC based login) this would be speculation at this point.

I will add a clarification about this intention to the MSC.

Author: Clarified in ad719369197fda5459d55759949932503f25f529

The sequence diagram for steps 4 and 5 is as follows:

```mermaid
sequenceDiagram
    participant E as Existing device <br>already signed in
    participant UA as Web Browser
    participant N as New device <br>wanting to sign in
    participant HS as Homeserver

    rect rgba(0,255,0, 0.1)

        E->>HS: GET /_matrix/client/v3/devices/{device_id}
        alt device already exists
            HS->>E: 200 OK
            E->>N: SecureSend({ "type": "m.login.failure", "reason": "device_already_exists" })
        else device not found
            HS->>E: 404 Not Found
        end
        par
            E->>N: SecureSend({"type":"m.login.protocol_accepted"})
        note over N: 4) New device polls the homeserver awaiting the outcome as per RFC 8628 / MSC4341
            loop Poll for result at interval <interval> seconds
                N->>HS: POST /token client_id=xyz<br>&grant_type=urn:ietf:params:oauth:grant-type:device_code<br>&device_code=XYZ
                alt pending
                    HS-->>N: 400 Bad Request {"error": "authorization_pending"}
                else granted
                    HS-->>N: 200 OK {"access_token": "...", "token_type": "Bearer", ...}

>                     N->>E: SecureSend({ "type": "m.login.success" })

Author: In step 4 above it says:

> - If the user consents in the next step then the new device will receive an `access_token` and `refresh_token` etc. as
normal as per [MSC4341].

The is also a note above which says "as per RFC 8628 / MSC4341".

Mitigations that are included in this proposal:

- the low maximum payload size (4KB)
- payload is restricted to string
- the rendezvous session should be short-lived
- use of `Sec-Fetch-*` headers to not return payload content when browser has navigated to the session URL

### QR code format

Once a device creates the rendezvous session it then generates a QR code that contains sufficient information for the
scanning device to locate the rendezvous session.

It is proposed that the QR code format that is currently used in the Client-Server API for
[device verification](https://spec.matrix.org/v1.16/client-server-api/#qr-code-format) be extended to be more general
purpose and accommodate this new use case.

The "QR code verification mode" would be changed to be something more general like "QR code mode".

We then define two new modes to allow for the two login device dispositions: the new device wishing to login; an existing
device wishing to facilitate the login of the new device;

The QR codes to be displayed and scanned using this format will encode binary strings in the general form:

- the ASCII string `MATRIX`

> - one byte indicating the QR code version (must be `0x02`)

Reviewer 2: What is the meaning and purpose of this field? What does the value 0x02 represent, i.e. what practical insight do you gain upon reading the value 0x02 in a message?

- one byte indicating the QR code mode. Should be one of the following values:
  - `0x03` a new device wishing to login and self-verify
  - `0x04` an existing device wishing to facilitate the login of a new device and self-verify that other device
- the ephemeral Curve25519 public key, as 32 bytes
- the rendezvous session ID encoded as:
  - two bytes in network byte order (big-endian) indicating the length in bytes of the rendezvous session ID as a UTF-8
  string
  - the rendezvous session ID as a UTF-8 string
- the [server name](https://spec.matrix.org/v1.15/appendices/#server-name) of the homeserver encoded as:
  - two bytes in network byte order (big-endian) indicating the length in bytes of the server name as a UTF-8 string
  - the server name as a UTF-8 string

If a new version of this QR sign in capability is needed in future (perhaps with updated secure channel protocol) then

> additional "mode" values can be added.

Reviewer 2: This feels like a hack. The modes were meant to encode directionality of the two participants, so it IMO makes little sense to add additional ones.

It can certainly be made to work for this particular case, but I'm beginning to wonder why this MSC does not have a single protocol versioning field. In light of https://github.com/matrix-org/matrix-spec-proposals/pull/4108/files#r2375468437, I believe the QR code version field should be clarified to be this missing protocol version field for the protocol described by this MSC.

4. **Existing device checks device_id and accepts protocol to use**

On receipt of the `m.login.protocol` message above, and having completed step 7 of the secure channel establishment, the
existing device then asserts that there is no existing device corresponding to the `device_id` from the
`m.login.protocol` message.

It does so by calling [GET /_matrix/client/v3/devices/<device_id>](https://spec.matrix.org/v1.9/client-server-api/#get_matrixclientv3devicesdeviceid)
and expecting to receive an HTTP 404 response.

If the device already exists then the login request should be rejected with an `m.login.failure` and reason `device_already_exists`:

*Existing device => New device via secure channel*

```json
{
    "type": "m.login.failure",
    "reason": "device_already_exists"
}
```

If no existing device was found then the existing device opens the `verification_uri_complete` - falling back to the
`verification_uri`, if `verification_uri_complete` isn't present - in a system browser.

Ideally this is in a trusted/secure environment where the cookie jar and password manager features are available. e.g.
on iOS this could be a `ASWebAuthenticationSession`

> The existing device then sends an acknowledgement message to let the other device know that the consent process is in progress:

Reviewer 9: When implementing this in matrix-rust-sdk, @poljar raised the question of whether the order of opening the browser on the existing device and sending the `m.login.protocol_accepted` message to the new device matters. Should we first send the acknowledgment so that the new device can start polling and only then open the browser to let the user consent?

I _think_ it doesn't make a huge difference but wanted to double check it here.

(Full discussion in https://github.com/matrix-org/matrix-rust-sdk/pull/5801#discussion_r2469444276)

Author: I have some recollection on this topic: 

On receipt of the verification URI the existing device might not be able to open it. Perhaps the operating system denies the request, or it turns out that the isn't a system browser available.

As such I *think* the intention was that `m.login.protocol_accepted` is only sent after the device knows that it can proceed with showing the URI.

But the MSC doesn't say this, and nor does the seem to be a suitable `m.login.failure` reason code defined.

If this was defined then I don't think it matters if the new device starts polling before it has received the `m.login.protocol_accepted`. But, it would need to stop and do something sensible if it received a `m.login.failure` instead of `m.login.protocol_accepted`.

The "QR code version" would be repurposed to be a "QR code type" and used as the way to distinguish the format of the
subsequent data.

The existing cross verification code would be type `0x02`.  I suspect that type `0x01` and `0x00` might correspond to
earlier iterations of the cross signing flow and so might want to be "reserved".

This proposal then adds a new type `0x03`.

The QR codes to be displayed and scanned using this format will encode binary strings in the general form:

- the ASCII string `MATRIX`
- one byte indicating the QR code type: `0x03` which identifies that the QR is part of this proposal
- one byte indicating the intent of the device generating the QR:
  - `0x00` a new device wishing to login and self-verify
  - `0x01` an existing device wishing to facilitate the login of a new device and self-verify that other device
- the ephemeral Curve25519 public key that will be used for [secure channel establishment](#establishment), as 32 bytes
- the rendezvous session ID encoded as:
  - two bytes in network byte order (big-endian) indicating the length in bytes of the rendezvous session ID as a UTF-8
  string
  - the rendezvous session ID as a UTF-8 string

> - the [server name](https://spec.matrix.org/v1.15/appendices/#server-name) of the homeserver encoded as:

Author: Reflecting on this some more: as discovery of the OAuth issuer is now done via the `/auth_metadata` [endpoint](https://spec.matrix.org/v1.16/client-server-api/#get_matrixclientv1auth_metadata) on the base URL (as opposed to from .well-known/matrix/client on the server name like it was when this proposal was originally written) I think it definitely makes sense to use the base URL. I will update the proposal to do this.

        N->>E: New device informs existing device that it has completed authentication<br><= m.login.success
        E->>HS: Existing device checks that new device has complete authentication
        HS-->>E: Homeserver confirms new device has completed authentication
        E->>N: Existing device sends new device the E2EE secrets<br>=>m.login.secrets

        note over N: New device stores secrets and verifies itself
        N->>HS: New device uploads device verification signature to homeserver
    end
```

### Login via OAuth 2.0 Device Authorization Grant from MSC4341

In this section the sequence of steps depends on whether the new device generated or scanned the QR code from [MSC4388].

For example, in the case that the new device scanned the QR code it is the first to do a `SecureSend` whereas if the new
device generated the QR then the existing device is the first to do a `SecureSend`.

Unfortunately, this can make it hard to read what is going on. Sequence diagrams are included for both variants after
the steps are described.

We use the `SecureSend` and `SecureReceive` operations from [MSC4388] which are sent via the out-of-band channel.

> 1. **Homeserver discovery**

Reviewer 10: These get rendered as single-element `<ol>` lists, which I'm finding surprisingly visually distracting.

Assuming we don't want to make the whole proposal into an actual list, maybe we should just reach for the fourth level of heading?

```suggestion
#### 1. Homeserver discovery
```

Author: Done in 754828108de636ffc23ef6e8a56ff8ede3c81159

The new device needs to know which homeserver it will be authenticating with.

In the case that the new device scanned the QR code then the [base URL]
of the Matrix homeserver can be taken from the QR code and the new device proceeds to step 2 immediately.

Otherwise the new device waits to be informed by receiving an `m.login.protocols` message from the existing device.

> The existing device would need to determine which "protocols" are available for the new device to use.

Reviewer 10: ```suggestion
The existing device would need to determine which "login protocols" are available for the new device to use.
```

maybe?

Currently this could only be `device_authorization_grant` meaning the homeserver supports the

> `urn:ietf:params:oauth:grant-type:device_code` grant type.

Reviewer 10: A general comment on the text of the MSC: I don't find these single-sentence paragraphs particularly easy to read. Can we maybe combine related ideas into paragraphs?

Author: I have attempted to strike a better balance in eeff6bb864b133b813454d41b52f44a9929fd519

Please say if you think even more is needed.

> If it is available then the existing device informs the new device by sending the `m.login.protocols` message with the

Reviewer 10: Is "it" the device_code grant type?

```suggestion
If that grant type is available, then the existing device informs the new device by sending the `m.login.protocols` message with the
```

Reviewer 10: What happens if that grant type is *not* available? Does the existing device just not send the `m.login.protocols` message?

Reviewer 10: > What happens if that grant type is _not_ available? Does the existing device just not send the `m.login.protocols` message?

Maybe the existing device should just not offer to show/scan the QR code? Would be good to be explicit.

Author: Addressed in e04dbc9626eb19601a4cd601395064d6a4503dba

homeserver specified:

*Existing device => New device via secure channel*

```json
{
    "type": "m.login.protocols",
    "protocols": ["device_authorization_grant"],
    "base_url": "https://synapse-oidc.lab.element.dev"
}
```

2. **New device checks if it can use an available protocol**

> The existing device then undertakes steps to determine if it is able to work with the homeserver.

Reviewer 10: ```suggestion
The new device then undertakes steps to determine if it is able to work with the homeserver.
```

I think?

- checks that the homeserver has the OAuth 2.0 API available by [`GET /_matrix/client/v1/auth_metadata`](https://spec.matrix.org/v1.15/client-server-api/#server-metadata-discovery) on the homeserver [base URL]

*New device => Homeserver via HTTP*

```http
GET /_matrix/client/v1/auth_metadata HTTP/1.1
Host: synapse-oidc.lab.element.dev
Accept: application/json
```

With response like:

```http
200 OK
Content-Type: application/json

{
    "issuer": "https://auth-oidc.lab.element.dev/",
    "authorization_endpoint": "https://auth-oidc.lab.element.dev/authorize",
    "token_endpoint": "https://auth-oidc.lab.element.dev/oauth2/token",
    "jwks_uri": "https://auth-oidc.lab.element.dev/oauth2/keys.json",
    "registration_endpoint": "https://auth-oidc.lab.element.dev/oauth2/registration",
    "scopes_supported": ["openid", "email"],
    "response_types_supported": [...],
    "response_modes_supported": [...],
    "grant_types_supported": [
        "authorization_code",
        "refresh_token",
        "client_credentials",
        "urn:ietf:params:oauth:grant-type:device_code"
    ],
    ...
    "device_authorization_endpoint": "https://auth-oidc.lab.element.dev/oauth2/device"
}

> ```

Reviewer 10: please indent all this by two spaces to avoid interrupting the bullet list.

Likewise for the other bullets below.

Author: Done in b6dc96c593979370e7d93a9d6fe67c5fcc5b9fc5

- either does Dynamic Client Registration as per the existing [spec](https://spec.matrix.org/v1.15/client-server-api/#client-registration)
or uses a static `client_id`. We will use `my_client_id` as an example `client_id`.

- sends a [RFC8628 Device Authorization Request](https://datatracker.ietf.org/doc/html/rfc8628#section-3.1) to the homeserver
using the `device_authorization_endpoint` as described by [MSC4341]:

*New device => Homeserver via HTTP*

```http
POST /oauth2/device HTTP/1.1
Host: auth-oidc.lab.element.dev
Content-Type: application/x-www-form-urlencoded

client_id=my_client_id&scope=openid%20urn%3Amatrix%3Aclient%3Aapi%3A%2A%20urn%3Amatrix%3Aclient%3Adevice%3AABCDEGH
```

With response like:

```http
200 OK
Content-Type: application/json

{
    "device_code": "GmRhmhcxhwAzkoEqiMEg_DnyEysNkuNhszIySk9eS",
    "user_code": "123456",
    "verification_uri": "https://auth-oidc.lab.element.dev/link",
    "verification_uri_complete": "https://auth-oidc.lab.element.dev/link?code=123456",
    "expires_in": 1800,
    "interval": 5
}
```

- parses the [Device Authorization Response](https://datatracker.ietf.org/doc/html/rfc8628#section-3.2) above

> At this point the new device knows that, subject to the user consenting, it should be able to complete the login

Reviewer 10: ```suggestion
At this point the new device knows that, subject to the user consenting, it should be able to complete the login.
```

The new device sends the `verification_uri` and, if present, the `verification_uri_complete` over to the existing device and
indicates that it wants to use protocol `device_authorization_grant` and that it will be authenticating as the Matrix
device with ID `device_id` (i.e. it will be requesting the [OAuth 2.0 API scope](https://spec.matrix.org/v1.16/client-server-api/#login-flow)
containing the specified device ID):

*New device => Existing device via secure channel*

```json
{
    "type": "m.login.protocol",
    "protocol": "device_authorization_grant",
    "device_authorization_grant": {
        "verification_uri": "https://auth-oidc.lab.element.dev/link",
        "verification_uri_complete": "https://auth-oidc.lab.element.dev/link?code=123456"
    },
    "device_id": "ABCDEFGH"
}
```

The sequence for steps 1 to 3 is as follows: (the sequence depending on which device  has scanned the code varies for readability)

_New device scanned QR code:_

```mermaid
sequenceDiagram
    title: Variant: New device scanned QR code
    participant E as Existing device <br>already signed in
    participant N as New device <br>wanting to sign in
    participant HS as Homeserver

>     rect rgba(255,0,0, 0.1)

Reviewer 10: is there a significance to the colour coding? Do the colours carry over between sequence diagrams?

Author: Yes, the idea was to highlight the difference between the two variants.

I've added a note to explain this in baf8b3f0d5319a0fb812e79270e61ce265dff9fa.

In 09d03a8f17e0a4bcde2ec0d0d58da95e66cb32ce I have also removed the background from the other diagrams where it adds nothing.

```json
{
    "type": "m.login.failure",
    "reason": "unable_to_open_verification_uri"
}
```

If the user denied permission to open the browser then instead the existing device sends an `m.login.failure` with
reason `user_cancelled`:

*Existing device => New device via secure channel*

```json
{
    "type": "m.login.failure",
    "reason": "user_cancelled"
}
```

5. **User is asked by homeserver to consent on existing device**

The user is then prompted to consent by the homeserver. They may be prompted to undertake additional actions by the
homeserver such as 2FA, but this is all handled within the browser.

Note that the existing device does not see the new access token. This is one of the benefits of the OAuth 2.0 API.

6. **New device waits for approval from homeserver**

In parallel to step 5, on receipt of the `m.login.protocol_accepted` message the new device:

- In accordance with [RFC8628](https://datatracker.ietf.org/doc/html/rfc8628#section-3.3.1) the new device must display
the `user_code` in order that the user can confirm it on the homeserver if required.
- The new device then starts to poll the homeserver by making

> [Device Access Token Requests](https://datatracker.ietf.org/doc/html/rfc8628#section-3.4) using the interval and bounded

Reviewer 10: which interval?

Author: Clarified in b6dc96c593979370e7d93a9d6fe67c5fcc5b9fc5. It's from the Device Authorization Response.

by `expires_in`.

> The above is as per [MSC4341].

Reviewer 10: Now that MSC4341 is canon, it might be nice to point to https://spec.matrix.org/v1.18/client-server-api/#device-authorization-flow instead.

Reviewer 10: Likewise other mentions of MSC4341

Author: Done in 76a9baec8bb2cf1f0eac030ce6852d4f421c7d4a

- The backup recovery key and the currently used backup version.

This is achieved as following:

1. **Existing device confirms that the new device has indeed logged in successfully**

On receipt of an `m.login.success` message the existing device queries the homeserver to check that there is a device online

with the corresponding device_id (from the `m.login.protocol` message).

It does so by calling [GET /_matrix/client/v3/devices/<device_id>](https://spec.matrix.org/v1.9/client-server-api/#get_matrixclientv3devicesdeviceid)
and expecting to receive an HTTP 200 response.

If the device isn't immediately visible it can repeat the `GET` request for up to, say, 10 seconds to allow for any latency.

If no device is found then the process should be stopped.

2. **Existing device shares secrets with new device**

The existing device sends a `m.login.secrets` message via the secure channel:

```json
{
    "type": "m.login.secrets",
    "cross_signing": {
        "master_key": "$base64_of_the_key",
        "self_signing_key": "$base64_of_the_key",
        "user_signing_key": "$base64_of_the_key"
    },
    "backup": {
        "algorithm": "foobar",

>         "key": "$base64_of_the_backup_recovery_key",

Reviewer 10: We don't use the term "recovery" in spec (https://spec.matrix.org/v1.18/client-server-api/#decryption-key)

```suggestion
        "key": "$base64_of_the_backup_decryption_key",
```

        "backup_version": "version_string"
    }
}
```

3. **New device cross-signs itself and uploads device keys**

> On receipt of the `m.login.secrets` message the new device can store the secrets locally

Reviewer 10: ```suggestion
On receipt of the `m.login.secrets` message the new device can store the secrets locally.
```

The new device can then generate the cross-signing signature for itself.

It can then use a single request to upload the device keys and cross signing signature. This removes the chance of other
devices seeing the new device as unverified, incorrectly prompting the user to verify the already verified device.

> The request would look just like any other `/keys/upload` request, it would just include one additional signature, the

Reviewer 10: ```suggestion
The request would look just like any other `/keys/upload` request; it would just include one additional signature, the
```

            "ed25519:SGKMSRAGBF": "b8gROFh+UIHLD/obY0+IlxoWiGtYVhKdqixvw4QHcN8"
        },
        "signatures": {
            "@testing_35:morpheus.localhost": {
                "ed25519:SGKMSRAGBF": "ziHEUIsHnrYBH4CqYpN1JC/ex3t4VG3zvo16D8ORqN6yAErpsKsnd/5LDdZERIOB1MGffKGfCL6ny5V7rT9FCQ",
                "ed25519:bkYgAVUNqvuyy8b1w09utJNJxBvK3hZB65xxoLPVzFol": "p257k0tfPF98OIDuXnFSJS2DmVlxO4sgTHdF41DTdZBCpTZfPwok6iASo3xMRKdyy3WMEgkQ6lzhEyRKKZBGBQ"
            }
        },
        "user_id": "@testing_35:morpheus.localhost"
    }
}
```

The sequence diagram for this would look as follows:

(for readability a pair of `SecureSend,SecureReceive` operations via the Homeserver is represented by a single
`SecureSendReceive` between the two devices)

```mermaid
sequenceDiagram
    participant E as Existing device <br>already signed in
    participant N as New device <br>wanting to sign in
    participant HS as Homeserver

    rect rgba(0,255,0, 0.1)
            rect rgb(191, 223, 255)
note over N,E: This step is duplicated from the previous section for readability
              N-->>+E: SecureSendReceive({ "type": "m.login.success" })

>             end

Reviewer 10: Alanis Morissette would like a word

<img width="1236" height="252" alt="Image" src="https://github.com/user-attachments/assets/1afc5cf1-a3ea-4f8f-8e14-1ed163d02b21" />

Author: To be fair, this how it renders for me at the moment:

<img width="622" height="144" alt="Image" src="https://github.com/user-attachments/assets/c72cf7ce-fb17-4192-9272-a1462975a20b" />

More generally the mermaid rendering is very unreadable in dark mode. I'm not sure why it isn't more sensible.

Author: I've removed the iron note in 09d03a8f17e0a4bcde2ec0d0d58da95e66cb32ce as I don't think it is actually needed.

I've checked how the diagrams are rendered depending on GitHub theme. What has caught me out is that there is a [bug](https://github.com/orgs/community/discussions/194348) in the mermaid rendering where it doesn't respect the system theme.

If you have the GitHub theme explicitly set to dark then it renders just fine. Light more is always fine.

Example:

```json
{
    "type": "m.login.secrets",
    "cross_signing": {
        "master_key": "$base64_of_the_key",
        "self_signing_key": "$base64_of_the_key",
        "user_signing_key": "$base64_of_the_key"
    },
    "backup": {
        "algorithm": "foobar",
        "key": "$base64_of_the_backup_recovery_key",
        "backup_version": "version_string"
    }
}
```

## Discoverability of the capability

Before offering this capability it would make sense that the device can check the availability of the feature.

Where the homeserver is known:

1. Check that the homeserver is using the OAuth 2.0 API using [server metadata discovery](https://spec.matrix.org/v1.15/client-server-api/#server-metadata-discovery)
1. Check that the Device Authorization Grant is available as per [MSC4341]
1. Check if the homeserver has a rendezvous session API available by attempting a POST to the create rendezvous endpoint
   from [MSC4388].

> For a new device it would need to know the homeserver ahead of time in order to do these checks.

Reviewer 10: What does this mean in practice? AIUI, the new device *doesn't* know the homeserver ahead of time, so can't do the checks?

Author: I think I have clarified this in 7239d2b7d259fac24405e5da2474b81e341150c0

Additionally the new device needs to either have an existing (i.e. static) OAuth 2.0 client registered with the homeserver
already, or the homeserver must support and allow dynamic client registration as described in the [spec](https://spec.matrix.org/v1.15/client-server-api/#client-registration).

The feature is also only available where a user has cross-signing set up and the existing device to be used has the

> Master Signing Key, Self Signing Key and User Signing Key stored locally so that they can be shared with the new device.

Reviewer 10: Is this strictly correct? it seems like the login part would still work?

Author: The login part *could* work, but it has been deliberately excluded from the scope of this MSC.

I have attempted to clarify by adding an alternative in f9106ceb943c7ef387dd20c1fbfb9f484e51a9f9

> Master Signing Key, Self Signing Key and User Signing Key stored locally so that they can be shared with the new device.

Reviewer 10: Shouldn't this whole thing form part of the main proposal, before "Login via OAuth 2.0 Device Authorization Grant from MSC4341"?

Author: Maybe... I've moved it up in 7239d2b7d259fac24405e5da2474b81e341150c0 and expanded it to more explicitly cover the different scenarios.

## Potential issues

This proposal adds new functionality it is not anticipated that it would conflict with other existing features.

Although this is providing a new authentication mechanism to Matrix it builds on top of the well established OAuth
Device Authorization Grant.

Because the cryptograpgic identity that is used for end-to-end encryption is being shared it is particularly important
to ensure that new attack vectors are not opened up.

A possible source of issues is the size and complexity of the proposal.

Please also see the potential issues from the dependent MSCs.

## Alternatives

### Alternative method of secret sharing

Instead of the existing device sharing the secrets bundle instead the existing device could cross-sign the new device
and then use to-device messaging for sharing the secrets.

For:

- You re-use existing secret sharing

Against:

- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.

> - Takes several round-trips for the secrets to be be shared which will add latency to the overall flow

Reviewer 10: ```suggestion
- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow.
```

- The backup cannot be immediately enabled since we received the backup version as well, something the `m.secret.send`

> mechanism does not offer.

Reviewer 10: Is this trying to say that we need to wait for a `GET /room_keys` request to the homeserver?

```suggestion
- Key backup upload cannot be enabled until we make a `GET /room_keys` request to the homeserver, in order to receive the receive the key backup version.
```

TBH I'm not sure that's a bad thing: it gives us a chance to ensure that the decryption key we were sent matches the public key for the backup version.

(Also, if all we want is the backup version, we could fix that in secret-sharing too, so I don't think this is an effective argument against using secret sharing for QR code login)

Possibly I'm missing something?

Reviewer 3: Yes, I think it's trying to say this.

> TBH I'm not sure that's a bad thing: it gives us a chance to ensure that the decryption key we were sent matches the public key for the backup version.

I would argue that due to the somewhat synchronous nature of QR code login, meaning you are right now talking to your own trusted device, you're not required to do this check in this instance.

> (Also, if all we want is the backup version, we could fix that in secret-sharing too, so I don't think this is an effective argument against using secret sharing for QR code login)

The backup version wouldn't be as useful in the secret-sharing scenario. Secret-sharing travels over to-device messaging which might not arrive immediately. As such double checking that the version is correct becomes a necessity.

That being said, I think there are enough other arguments in this section. I think we can just remove this one if you remain unconvinced of this argument.

Reviewer 10: Resolving in favour of a [new thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r3504594875)

  }
  ```

- parses the [Device Authorization Response] above

At this point the new device knows that, subject to the user consenting, it should be able to complete the login.

#### 3. New device informs existing device that it wants to use the `device_authorization_grant`

The new device sends the `verification_uri` and, if present, the `verification_uri_complete` over to the existing device and
indicates that it wants to use protocol `device_authorization_grant` and that it will be authenticating as the Matrix
device with ID `device_id` (i.e. it will be requesting the [OAuth 2.0 API scope](https://spec.matrix.org/v1.16/client-server-api/#login-flow)
containing the specified device ID):

*New device => Existing device via secure channel*

```json
{
    "type": "m.login.protocol",
    "protocol": "device_authorization_grant",
    "device_authorization_grant": {
        "verification_uri": "https://auth-oidc.lab.element.dev/link",
        "verification_uri_complete": "https://auth-oidc.lab.element.dev/link?code=123456"
    },
    "device_id": "ABCDEFGH"
}
```

The sequence for steps 1 to 3 is as follows: (the sequence depending on which device  has scanned the code varies for

> readability. The steps that are the same have a green background, the steps that are different have a red background)

Reviewer 10: > the sequence depending on which device  has scanned the code varies for readability.

I am struggling to parse this sentence.

Author: In d16d06cca21eea89b7fb05ff3d5f149ea53423ce I have reworded so that it hopefully actually makes sense now.

_New device scanned QR code:_

```mermaid
sequenceDiagram

>     title: Variant: New device scanned QR code

Reviewer 10: This title repeats the heading just above the diagram, so I don't think it adds any clarity. Suggest getting rid of one of the titles.

Author: Done in a0488e5a9c6d4aea7ba372c3cfef61b3e5ac0e8e

        note over N: New device displays checkmark and CheckCode from MSC4388
        note over E: Existing device waits for user to enter CheckCode<br>and confirm secure channel from MSC4388 step 7
    end

    rect rgba(0,255,0, 0.1)
        HS->>E: SecureReceive() => {"type": "m.login.protocol", "protocol": "device_authorization_grant",<br> "device_authorization_grant":{<br>"verification_uri_complete": "https://id.matrix.org/device/abcde",<br>"verification_uri": ...}, "device_id": "ABCDEFGH"}
    end

    rect rgba(255,0,0, 0.1)
    # alt if New device scanned QR code
        note over E: If user entered correct CheckCode<br>and confirms checkmark then existing device now trusts the channel
    end

    rect rgba(0,255,0, 0.1)
    note over E: Existing device checks that requested protocol is supported

    alt if requested protocol is not valid
        E->>HS: SecureSend({"type":"m.login.failure", "reason":"unsupported_protocol",<br>"homeserver": "matrix.org"})
        HS->>N: SecureReceive({"type":"m.login.failure", "reason":"unsupported_protocol",<br>"homeserver": "matrix.org"})
    end
    end
```

_Existing device scanned QR code:_

```mermaid
sequenceDiagram

>     title: Variant: Existing device scanned QR code

Reviewer 10: likewise, repetition.

Author: Done in a0488e5a9c6d4aea7ba372c3cfef61b3e5ac0e8e

    }
}
```

## Potential issues

This proposal adds new functionality and is not anticipated to conflict with other existing features. Although it
provides a new authentication mechanism, it builds on the well-established OAuth Device Authorization Grant.

Because the cryptographic identity used for end-to-end encryption is being shared, it is particularly important to
ensure that new attack vectors are not opened up. A possible source of issues is the size and complexity of the
proposal. Please also see the potential issues from the dependent MSCs.

## Alternatives

### Alternative method of secret sharing

Instead of sharing a secrets bundle directly, the existing device could cross-sign the new device and then use
to-device messaging for sharing the secrets.

For:

- You re-use existing secret sharing

Against:

- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.
- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow.
- The backup cannot be immediately enabled since we received the backup version as well, something the `m.secret.send`

>   mechanism does not offer.

Reviewer 10: This is still somewhat confusingly worded. (See [previous thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r3305677813))

Following @poljar's comments on that thread, maybe we can just note that changes *could* be made to the `m.secret.send` mechanism if this was the only problem:

```suggestion
- Key backup upload cannot be enabled until we make a `GET /room_keys` request to the homeserver, in order to receive the receive the key backup version (though the `m.secret.send` mechanism could potentially be modified to provide this information).
```

  or uses a static `client_id`. We will use `my_client_id` as an example `client_id`.

- sends a [RFC8628 Device Authorization Request](https://datatracker.ietf.org/doc/html/rfc8628#section-3.1) to the homeserver

>   using the `device_authorization_endpoint` as described by the [device authorization grant]:

Reviewer 10: I think this was intended to be a link?

```suggestion
  using the `device_authorization_endpoint` as described by the [device authorization flow]:
```

###### Maximum duration of a rendezvous

> The rendezvous session needs to persist for the duration of the login. So a timeout such as 60 seconds should be adequate.

Reviewer 4: I think 60 seconds is very tight if you have a slow connection, not because of bandwidth but because of latency in establishing the various connections. I'd go for 180 seconds, which I have to admit is just a guesstimate.

Reviewer 6: I think defining a minimum duration instead of a maximum would be preferable

Reviewer 4: You want a maximum duration to avoid abuse of the service.

Reviewer 6: Yes, but maybe the maximum should be up to the implementation while the spec defines a minimum that should work in most cases

Reviewer 4: I think the spec should not lean towards vulnerable implementations. A maximum duration is the way to go if you ask me.

Author: I've incorporated a max and min into https://github.com/matrix-org/matrix-spec-proposals/commit/34ade3c1ac66916430cc7e20ad3e7bb0d2518a2a
